experimental/displaynames/src/displaynames.rs

```diff
@@ -494,148 +438,185 @@ impl LocaleDisplayNamesFormatter {
         }
         result
     }
-}
 
-fn get_locale_display_name<'a>(
-    locale: &Locale,
-    longest_matching_subtag: LongestMatchingSubtag,
-    locale_dn_formatter: &'a LocaleDisplayNamesFormatter,
-) -> &'a str {
-    let LocaleDisplayNamesFormatter {
-        options,
-        locale_data,
-        language_data,
-        ..
-    } = locale_dn_formatter;
-
-    let lang_id: LanguageIdentifier = match longest_matching_subtag {
-        LongestMatchingSubtag::LangRegion => (locale.id.language, None, locale.id.region).into(),
-        LongestMatchingSubtag::LangScript => (locale.id.language, locale.id.script, None).into(),
-        LongestMatchingSubtag::Lang => locale.id.language.into(),
-    };
-
-    // Check if the key exists in the locale_data first.
-    // Example: "en_GB", "nl_BE".
-    let mut ldn = match options.style {
-        Some(Style::Short) => locale_data
-            .get()
-            .short_names
-            .get_by(|uvstr| lang_id.strict_cmp(uvstr).reverse()),
-        Some(Style::Long) => locale_data
-            .get()
-            .long_names
-            .get_by(|uvstr| lang_id.strict_cmp(uvstr).reverse()),
-        Some(Style::Menu) => locale_data
-            .get()
-            .menu_names
-            .get_by(|uvstr| lang_id.strict_cmp(uvstr).reverse()),
-        _ => None,
+    /// For a given locale and the data, find the longest prefix of the string that exists as a key in the CLDR locale data.
+    pub fn find_longest_matching_subtag(&self, locale: &Locale) -> LanguageIdentifier {
+        let LocaleDisplayNamesFormatter { locale_data, .. } = self;
```

Comment:
nit: I personally would just use `self.locale_data` everywhere instead of doing this, but up to you

Reply:
Done